### PR TITLE
Bugfix for computations of `δψ` and supercells

### PR DESCRIPTION
**Pull request by epolack**
Fix to propagate temperature, smearing and Fermi level with supercell method. Also creation of an `Hamiltonian` in `scfres` to enable the use of response functions.

### CODE REVIEW

File: src/supercell.jl [hunk 1]
```diff
@@ -32,9 +32,13 @@ function cell_to_supercell(basis::PlaneWaveBasis)
     supercell_size = basis.kgrid
     supercell = create_supercell(model.lattice, model.atoms, model.positions, supercell_size)
     supercell_fft_size = basis.fft_size .* supercell_size
+    # Keep some values unchanged from the cell model
+    cell_kwargs = (; terms=model.term_types, model.temperature, model.smearing, model.εF,
```
Comment (mfherbst) [inline src/supercell.jl:36]: model_kwargs ?

File: src/supercell.jl [hunk 2]
```diff
@@ -32,9 +32,13 @@ function cell_to_supercell(basis::PlaneWaveBasis)
...
+                   # Can be safely disabled; this has already been checked for basis.model
+                   disable_electrostatics_check=true)
     # Assemble new model and new basis
     model_supercell = Model(supercell.lattice, supercell.atoms, supercell.positions;
-                            terms=model.term_types, symmetries=false)
+                            cell_kwargs..., symmetries=false)
```
Comment (mfherbst) [inline src/supercell.jl:41]: There is even more. Just use the pattern of 
https://github.com/JuliaMolSim/DFTK.jl/blob/master/src/postprocess/stresses.jl
Btw there the εF is not passed, maybe one should do a pass to fix the stresses and equivalent places where the model is reconstructed elsewhere (e.g. serialisation, forward-diff are places that come to my mind.

File: src/supercell.jl [hunk 3]
```diff
@@ -115,7 +119,7 @@ function cell_to_supercell(scfres::NamedTuple)
     energies_supercell = [n_unit_cells * v for v in values(scfres_unfold.energies)]
```
Comment (mfherbst) [inline src/supercell.jl:119]: I would make the ham up here using `energies_hamiltonian`. This is called by `Hamiltonian( )` anyway.

File: test/supercell.jl [hunk 4]
```diff
@@ -4,15 +4,30 @@ include("testcases.jl")
 
 if mpi_nprocs() == 1  # can't be bothered to convert the tests
 
+# Quick test to make sure temperature, smearing and Fermi level are correctly propagated
```
Comment (mfherbst) [inline test/supercell.jl:7]: Also test spin.

**Comment by LaurentVidal95** [comment]
> @LaurentVidal95 would that be okay? Especially for the `Hamiltonian` modification of `scfres`.

Hey yep it doesn't change anything on my side !

**Comment by antoine-levitt** [comment]
@gkemlin does this look OK?

**Comment by gkemlin** [comment]
It looks good to me. We indeed do not need `δHψ` for discarded states, good catch !

File: src/response/chi0.jl [hunk 5]
```diff
@@ -346,20 +346,25 @@ end
...
 
     ε_occ   = [eigenvalues[ik][maskk] for (ik, maskk) in enumerate(mask_occ)]
+    δHψ_occ = [δHψ[ik][:, maskk] for (ik, maskk) in enumerate(mask_occ)]
 
     occ_occ = [occ[ik][maskk] for (ik, maskk) in enumerate(mask_occ)]
 
-    # First we compute δoccupation and δεF
-    δocc, δεF = compute_δocc(basis, ψ_occ, occ_occ, εF, ε_occ, δHψ)
-    # Pad δoccupation
+    # First we compute masked δoccupation and δεF…
```
Comment (antoine-levitt) [inline src/response/chi0.jl:353]: delta occupation and eps f for the occupied orbitals (the mask is just the technical way we select the occupied orbitals), same below

**Review by antoine-levitt — COMMENTED** [review]
Looks good, thanks!

**Comment by antoine-levitt** [comment]
LGTM!

File: src/show.jl [hunk 6]
```diff
@@ -29,11 +29,14 @@ function Base.show(io::IO, ::MIME"text/plain", model::Model)
     end
 
     println(io)
-    showfieldln(io, "num. electrons",    model.n_electrons)
-    showfieldln(io, "spin polarization", model.spin_polarization)
-    showfieldln(io, "temperature",       @sprintf "%.5g Ha" model.temperature)
+    showfieldln(io, "num. electrons",        model.n_electrons)
```
Comment (mfherbst) [inline src/show.jl:32]: Should this not have an `isnothing` as well?

File: test/serialisation.jl [hunk 7]
```diff
@@ -36,10 +38,10 @@ end
...
 
 @testset "Test checkpointing" begin
-    O = ElementPsp(o2molecule.atnum, psp=load_psp("hgh/pbe/O-q6.hgh"))
-    model = model_PBE(o2molecule.lattice, [O, O], o2molecule.positions;
-                      temperature=0.02, smearing=Smearing.Gaussian(),
-                      magnetic_moments=[1., 1.], symmetries=false)
+    model = model_PBE(o2molecule.lattice, o2molecule.atoms, o2molecule.positions;
+                      temperature=0.02, smearing=Smearing.Gaussian(), εF=0.5,
+                      magnetic_moments=[1., 1.], symmetries=false,
+                      disable_electrostatics_check=true)
```
Comment (mfherbst) [inline test/serialisation.jl:44]: ok, but don't we ideally have both checks?

File: test/supercell.jl [hunk 8]
```diff
@@ -4,15 +4,31 @@ include("testcases.jl")
...
+@testset "Supercell copy" begin
+    Ecut    = 4
+    kgrid   = [2, 1, 1]
+    # Parameters
+    model = model_LDA(magnesium.lattice, magnesium.atoms, magnesium.positions;
+                      magnesium.temperature, εF=0.5, spin_polarization=:spinless,
+                      disable_electrostatics_check=true)
+    basis = PlaneWaveBasis(model; Ecut, kgrid)
+    # SCF
```
Comment (mfherbst) [inline test/supercell.jl:16]: pointless comment

File: test/supercell.jl [hunk 9]
```diff
@@ -31,4 +47,42 @@ if mpi_nprocs() == 1  # can't be bothered to convert the tests
     @test norm(ρ_ref .- scfres_supercell_manual.ρ) < 10*tol
 end
 
+@testset "Supercell response" begin
+    Ecut    = 5.0
+    kgrid   = [2, 1, 1]
+    scf_kw  = (; is_converged=DFTK.ScfConvergenceDensity(1e-9),
+               determine_diagtol=DFTK.ScfDiagtol(diagtol_max=1e-9),
+               eigensolver=diag_full,
+               callback=identity)
```
Comment (mfherbst) [inline test/supercell.jl:56]: no need for linebreak

File: test/supercell.jl [hunk 10]
```diff
@@ -31,4 +47,42 @@ if mpi_nprocs() == 1  # can't be bothered to convert the tests
     @test norm(ρ_ref .- scfres_supercell_manual.ρ) < 10*tol
 end
 
+@testset "Supercell response" begin
+    Ecut    = 5.0
+    kgrid   = [2, 1, 1]
+    scf_kw  = (; is_converged=DFTK.ScfConvergenceDensity(1e-9),
+               determine_diagtol=DFTK.ScfDiagtol(diagtol_max=1e-9),
```
Comment (mfherbst) [inline test/supercell.jl:54]: Do you need that?

**Comment by mfherbst** [comment]
Indeed. I will do the rebase and get this PR merged once #800 is in.

**Comment by mfherbst** [comment]
CI breaks because of https://github.com/timholy/SnoopCompile.jl/issues/317

**Pull request merged**